File: ABC/351-400/374/d.cpp
#include <bits/stdc++.h>
using namespace std;
#define rep(i,n) for(int i=0;i<n;++i)
struct Line{
    int a,b,c,d;
    Line(int a=0,int b=0,int c=0,int d=0):a(a),b(b),c(c),d(d){}
    void input(){
        cin >> a >> b >> c >> d;
    }
};

int main(){
    int n,s,t;
    cin >> n >> s >> t;
    vector<Line> ls(n);
    rep(i,n){
        ls[i].input();
    }
    vector<int> is;
    rep(i,n) is.push_back(i);
    int n2 = 1<<n;
    auto dist = [&](double x1, double y1, double x2,double y2,bool on){
        double res = 0;
        res = (x1-x2)*(x1-x2) + (y1-y2)*(y1-y2);
        res = sqrt(res);
        if(on) return res/(double)t;
        else return res/(double)s;
    };
    double ans = 1e18;
    do{
        rep(i,n2){
            double si=0,sj=0;
            double now = 0;
            rep(j,n){
                if(i>>j&1){ //(A,B) -> (C,D)
                    now += dist(si,sj,ls[is[j]].a,ls[is[j]].b,false);
                    si = ls[is[j]].a;
                    sj = ls[is[j]].b;
                    now += dist(si,sj,ls[is[j]].c,ls[is[j]].d,true);
                    si = ls[is[j]].c;
                    sj = ls[is[j]].d;
                }else{
                    now += dist(si,sj,ls[is[j]].c,ls[is[j]].d,false);
                    si = ls[is[j]].c;
                    sj = ls[is[j]].d;
                    now += dist(si,sj,ls[is[j]].a,ls[is[j]].b,true);
                    si = ls[is[j]].a;
                    sj = ls[is[j]].b;
                }
            }
            ans = min(ans,now);
        }
    }while(next_permutation(is.begin(),is.end()));
    printf("%.10f\n",ans);
    return 0;
}

File: ABC/351-400/374/d.py
from itertools import permutations
import math
def main():
 n,s,t = map(int,input().split())
 a,b,c,d = [],[],[],[]
 for i in range(n):
  ai,bi,ci,di = map(int,input().split())
  a.append(ai)
  b.append(bi)
  c.append(ci)
  d.append(di)
  
 def dist(cx,cy,x,y,on):
   res = (cx-x)**2 + (cy-y)**2
   res = math.sqrt(res)
   if on:
     return res/t
   else:
     return res/s
 ans = 1e18
 for v in permutations(range(n)):
   for i in range(1<<n):
     now = 0
     ci,cj = 0,0 
     for j in range(n):
       if i>>j&1:
         now += dist(ci,cj,a[v[j]],b[v[j]],False)
         ci = a[v[j]]
         cj = b[v[j]]
         now += dist(ci,cj,c[v[j]],d[v[j]],True)
         ci = c[v[j]]
         cj = d[v[j]]
       else:
         now += dist(ci,cj,c[v[j]],d[v[j]],False)
         ci = c[v[j]]
         cj = d[v[j]]
         now += dist(ci,cj,a[v[j]],b[v[j]],True)
         ci = a[v[j]]
         cj = b[v[j]]
     ans = min(ans,now)
 print(ans)



  
if __name__ == '__main__':
  main()


